plugins/pi4j-plugin-linuxfs/src/main/java/com/pi4j/plugin/linuxfs/provider/spi/LinuxFsSpi.java

```diff
@@ -364,33 +357,62 @@ public int write(byte b) {
     public int write(byte[] data, int offset, int length) {
         Objects.checkFromIndexSize(offset, length, data.length);
 
+        int fullEntries = (length < SPI_BUFFSIZ) ? 1 : length / SPI_BUFFSIZ;
+        int partialEntries = (((fullEntries * SPI_BUFFSIZ) < length) ? 1 : 0);
+        int numberXferEntries = fullEntries + partialEntries;
+        spi_ioc_transfer[] transferArray = getTransferArray(numberXferEntries);
+        Memory[] peerArray = getPeerMem(numberXferEntries, SPI_BUFFSIZ) ;
         int start = offset;
-        while (start < data.length) {
-            PeerAccessibleMemory buf = new PeerAccessibleMemory(SPI_BUFFSIZ);
-            spi_ioc_transfer txEntry = new spi_ioc_transfer() ;
-            int end = Math.min(data.length, start + SPI_BUFFSIZ);
-            byte[] chunk = Arrays.copyOfRange(data, start, end);
-            buf.write(0, chunk, 0 , chunk.length);
+        int lastOffset = data.length + offset ;
+        int entryNum = 0;
+        while (start < lastOffset) {
+            //  PeerAccessibleMemory buf = new PeerAccessibleMemory(SPI_BUFFSIZ);
+            spi_ioc_transfer txEntry = transferArray[entryNum];
+            int end = Math.min(lastOffset, start + SPI_BUFFSIZ);
+            peerArray[entryNum].write(0, data, start, end - start);
             // set fields in transfer msg
-            txEntry.tx_buf = buf.getPeer();
+            txEntry.tx_buf = Memory.nativeValue(peerArray[entryNum]) ;
             txEntry.rx_buf = 0;
             txEntry.bits_per_word = BITS8;
             txEntry.speed_hz = config.baud();
             txEntry.delay_usecs = 0;
-            txEntry.len = chunk.length ;
-            txEntry.cs_change = 1 ;
-            int ret = libc.ioctl(fd, SPI_IOC_MESSAGE(1), txEntry);
-            logger.trace("[SPI::WRITE] <- Number bytes {} ", ret);
-            if (ret < 0) {
-                logger.error("Could not write SPI message. ret {}, error: {}", ret, Native.getLastError());
-                length = 0;
-            }
+            txEntry.len = end - start;
+            txEntry.cs_change = 0;
+            transferArray[entryNum] = txEntry;
+            entryNum++;
             start += SPI_BUFFSIZ;
         }
+        // last entry has chip select change  set to  deassert
+        transferArray[entryNum - 1].cs_change = 1;
+        int ret = libc.ioctl(fd, SPI_IOC_MESSAGE(numberXferEntries), transferArray);
+        logger.trace("[SPI::WRITE] <- Number bytes {} ", ret);
+
+        if (ret < 0) {
+            logger.error("Could not write SPI message. ret {}, error: {}", ret, Native.getLastError());
```

Comment:
I think we should discuss consistency across plugins / providers for this case separately:

Java only does this when running into a blocked read case -- errors are always reported as exceptions. 

And there the return value will never be 0

Moreover, writing 0 bytes could be a degenerated case where just nothing is written, and user code would need to be aware of this and distinguish the cases...